File: sdk/python/tipsy_ab_config/cache.py
"""Local NamespaceSnapshot cache with dual snapshot_seq semantics.

Mirrors the Go ``configCache`` exactly:

- Per-namespace immutable ``NamespaceSnapshot`` value objects.
- ``apply`` replaces the cached snapshot iff EITHER ``business_snapshot_seq``
  or ``experiment_snapshot_seq`` strictly advances (including the 0 → 1
  bootstrap case).
- Single ``threading.Lock`` guards the namespace map; readers grab the
  snapshot pointer lock-free and treat it as immutable.
- ``known_seqs`` builds the per-ns pair the SDK sends with each PullAll /
  Subscribe.

This module is import-light: no asyncio, no grpcio. It can be exercised from
synchronous code paths (Flask sync) the same way the Go side can be called
from any goroutine.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from threading import Lock
from typing import Dict, Iterable, List, Mapping, Optional, Tuple


@dataclass(frozen=True)
class KeyState:
    """SDK-local view of one config key.

    ``full_release_version`` is ``None`` when the key has no active full
    release; ``versions`` maps version_id → value (full release version + any
    versions abtest has flagged as potentially applicable).
    """

    full_release_version: Optional[int]
    versions: Mapping[int, str]


@dataclass(frozen=True)
class NamespaceSnapshot:
    """SDK-local view of one NamespaceSnapshot.

    Both seqs MUST be set; ``experiment_snapshot_seq == 0`` is the documented
    degraded marker (abtest unavailable, config fell back to "full-release
    only").
    """

    namespace: str
    business_snapshot_seq: int
    experiment_snapshot_seq: int
    keys: Mapping[str, KeyState]
    cached_bytes: int = 0


@dataclass
class ApplyResult:
    """Diagnostic record returned by ``ConfigCache.apply``."""

    replaced: bool = False
    business_moved: bool = False
    experiment_moved: bool = False


class ConfigCache:
    """Per-process namespace cache."""

    def __init__(self) -> None:
        self._lock = Lock()
        self._by_ns: Dict[str, NamespaceSnapshot] = {}

    # ---- public reads ----

    def snapshot(self, namespace: str) -> Optional[NamespaceSnapshot]:
        with self._lock:
            return self._by_ns.get(namespace)

    def full_release_version(self, namespace: str, key: str) -> Optional[int]:
        """Return the active full-release version_id for ``(ns, key)``.

        ``None`` on a cache miss (no snapshot, no such key, or the key has no
        full release set).  Callers MUST gate on ``is None``.
        """
        snap = self.snapshot(namespace)
        if snap is None:
            return None
        ks = snap.keys.get(key)
        if ks is None or ks.full_release_version is None:
            return None
        return ks.full_release_version

    def value_of(
        self, namespace: str, key: str, version_id: int
    ) -> Optional[str]:
        """Return the cached value for ``(ns, key, version_id)``.

        ``None`` on a cache miss.  Empty string / "[]" / "{}" are all valid
        values; callers MUST use ``is None`` to test for "no value" per
        design §10.5.
        """
        snap = self.snapshot(namespace)
        if snap is None:
            return None
        ks = snap.keys.get(key)
        if ks is None:
            return None
        return ks.versions.get(version_id)

    def list_namespaces(self) -> List[str]:
        with self._lock:
            return sorted(self._by_ns.keys())

    def known_seqs(
        self, namespaces: Iterable[str]
    ) -> Dict[str, Tuple[int, int]]:
        """Return per-ns (business_seq, experiment_seq) for the listed ns.

        Namespaces not yet cached are returned as (0, 0) so the server treats
        them as "send everything" per backend §4.1.2.
        """
        out: Dict[str, Tuple[int, int]] = {}
        with self._lock:
            for ns in namespaces:
                snap = self._by_ns.get(ns)
                if snap is None:
                    out[ns] = (0, 0)
                else:
                    out[ns] = (
                        snap.business_snapshot_seq,
                        snap.experiment_snapshot_seq,
                    )
        return out

    # ---- mutate ----

    def apply(self, pb_snapshot) -> ApplyResult:
        """Apply a protobuf NamespaceSnapshot.

        Replaces the cached snapshot iff either snapshot_seq is strictly
        greater than the cached value.  Returns an ``ApplyResult`` that
        records which seq moved so the caller can fire metrics.
        """
        result = ApplyResult()
        if pb_snapshot is None or not getattr(pb_snapshot, "namespace", ""):
            return result
        ns = pb_snapshot.namespace
        new_biz = int(pb_snapshot.business_snapshot_seq)
        new_exp = int(pb_snapshot.experiment_snapshot_seq)

        with self._lock:
            current = self._by_ns.get(ns)
            cur_biz = current.business_snapshot_seq if current else 0
            cur_exp = current.experiment_snapshot_seq if current else 0
            result.business_moved = new_biz > cur_biz
            result.experiment_moved = new_exp > cur_exp
            if current is not None and not (
                result.business_moved or result.experiment_moved
            ):
                return result

            keys: Dict[str, KeyState] = {}
            byte_size = 0
            for k in pb_snapshot.keys:
                if not k.key:
                    continue
                # `optional int64 full_release_version` — use HasField to
                # discriminate "unset" from "id = 0".
                if k.HasField("full_release_version"):
                    full_v: Optional[int] = int(k.full_release_version)
                else:
                    full_v = None
                versions: Dict[int, str] = {}
                for vid, val in k.versions.items():
                    versions[int(vid)] = val
                    byte_size += len(val)
                byte_size += len(k.key)
                keys[k.key] = KeyState(
                    full_release_version=full_v, versions=versions
                )
            self._by_ns[ns] = NamespaceSnapshot(
                namespace=ns,
                business_snapshot_seq=new_biz,
                experiment_snapshot_seq=new_exp,
                keys=keys,
                cached_bytes=byte_size,
            )
            result.replaced = True
        return result

File: sdk/go/tipsyabconfig/cache.go
package tipsyabconfig

import (
	"sort"
	"sync"

	configv1 "github.com/Lightspeed-Intelligence/tipsy-ab-config-sdk/api/gen/go/tipsy/config/v1"
)

// KeyState is the SDK-local mirror of configv1.KeyState. It is a value type
// embedded in NamespaceSnapshot so reads do not need pointer indirection.
type KeyState struct {
	// FullReleaseVersion is the version_id of the active full release for
	// this key, or 0 if none. We use 0 as the sentinel because proto3
	// `optional int64` is normalised here: the SDK only stores positive
	// version ids on the wire; the host caller never sees the raw struct.
	FullReleaseVersion int64
	// Versions maps version_id → value. Contains the full-release version
	// (if any) plus every version that abtest reports as "possibly
	// applicable" for the namespace.
	Versions map[int64]string
	// HasDynamicResolution mirrors the proto's `optional bool` presence
	// semantics: nil = field absent (old server), &false = explicit "pure
	// full-release, no gray/experiment", &true = needs abtest resolution.
	// GetConfig only takes the fast-path (skip abtest) when this is present
	// AND false; nil (old server) or true keeps the existing abtest path so a
	// new SDK against an old server never silently skips a live experiment.
	HasDynamicResolution *bool
}

// NamespaceSnapshot is the per-ns immutable view the cache hands out to
// readers. The dual snapshot_seq pair (business + experiment) is the freshness
// vector — either advancing triggers a cache replace.
type NamespaceSnapshot struct {
	Namespace             string
	BusinessSnapshotSeq   int64
	ExperimentSnapshotSeq int64
	Keys                  map[string]KeyState
}

// configCache is the per-process in-memory cache of NamespaceSnapshot.
//
// Concurrency model: many readers, one writer per namespace at a time. We use
// a sync.RWMutex on the namespace map; the per-ns snapshot pointer itself is
// atomic by virtue of the map slot write under the writer lock. Readers grab
// an RLock, fetch the pointer, then read its fields lock-free (the snapshot
// is treated as immutable once published).
type configCache struct {
	mu      sync.RWMutex
	byNS    map[string]*NamespaceSnapshot
	metrics *Metrics
}

func newConfigCache(m *Metrics) *configCache {
	return &configCache{byNS: make(map[string]*NamespaceSnapshot), metrics: m}
}

// snapshot returns the current snapshot for ns, or nil if absent. Safe for
// concurrent use; the returned pointer is treated as immutable.
func (c *configCache) snapshot(ns string) *NamespaceSnapshot {
	c.mu.RLock()
	s := c.byNS[ns]
	c.mu.RUnlock()
	return s
}

// fullReleaseVersion returns (versionID, true) when an active full-release
// version is present for (ns, key). It returns (0, false) on a miss; callers
// MUST NOT treat 0 as "no full release" — use the second return value.
func (c *configCache) fullReleaseVersion(ns, key string) (int64, bool) {
	s := c.snapshot(ns)
	if s == nil {
		return 0, false
	}
	ks, ok := s.Keys[key]
	if !ok || ks.FullReleaseVersion == 0 {
		return 0, false
	}
	return ks.FullReleaseVersion, true
}

// hasDynamicResolution reports the cached has_dynamic_resolution flag for
// (ns, key) with presence semantics. It returns (val, true) only when a
// snapshot for ns exists, the key is present, AND the field was explicitly set
// by the server. It returns (false, false) when the ns, key, or field is
// absent — i.e. present == false means "unknown / old server", NOT false.
// GetConfig MUST gate its fast-path on present == true && val == false so an
// absent field never triggers a skip.
func (c *configCache) hasDynamicResolution(ns, key string) (val bool, present bool) {
	s := c.snapshot(ns)
	if s == nil {
		return false, false
	}
	ks, ok := s.Keys[key]
	if !ok || ks.HasDynamicResolution == nil {
		return false, false
	}
	return *ks.HasDynamicResolution, true
}

// valueOf returns (value, true) when a specific (ns, key, versionID) is
// cached. It returns ("", false) on a cache miss. Per design §10.5 the empty
// string is a valid value; callers MUST gate on the second return value.
func (c *configCache) valueOf(ns, key string, versionID int64) (string, bool) {
	s := c.snapshot(ns)
	if s == nil {
		return "", false
	}
	ks, ok := s.Keys[key]
	if !ok {
		return "", false
	}
	v, ok := ks.Versions[versionID]
	return v, ok
}

// applyProto replaces the cached snapshot for ns iff either snapshot_seq
// advanced. It returns the (oldSnapshot, newSnapshot, replaced) triple so
// callers can fire metrics. Per design §4 either business OR experiment
// advancing — including 0 → 1 — triggers a replace.
func (c *configCache) applyProto(pb *configv1.NamespaceSnapshot) (replaced, businessMoved, experimentMoved bool) {
	if pb == nil || pb.Namespace == "" {
		return false, false, false
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	cur := c.byNS[pb.Namespace]

	newBiz := pb.BusinessSnapshotSeq
	newExp := pb.ExperimentSnapshotSeq

	var curBiz, curExp int64
	if cur != nil {
		curBiz = cur.BusinessSnapshotSeq
		curExp = cur.ExperimentSnapshotSeq
	}

	businessMoved = newBiz > curBiz
	experimentMoved = newExp > curExp
	if cur != nil && !businessMoved && !experimentMoved {
		return false, false, false
	}

	next := &NamespaceSnapshot{
		Namespace:             pb.Namespace,
		BusinessSnapshotSeq:   newBiz,
		ExperimentSnapshotSeq: newExp,
		Keys:                  make(map[string]KeyState, len(pb.Keys)),
	}
	var byteSize uint64
	for _, k := range pb.Keys {
		if k == nil || k.Key == "" {
			continue
		}
		ks := KeyState{Versions: make(map[int64]string, len(k.Versions))}
		if k.FullReleaseVersion != nil {
			ks.FullReleaseVersion = *k.FullReleaseVersion
		}
		// Preserve presence: copy the pointer's value into a fresh local
		// pointer iff set, leaving nil when the proto field is absent.
		if k.HasDynamicResolution != nil {
			hdr := *k.HasDynamicResolution
			ks.HasDynamicResolution = &hdr
		}
		for vid, val := range k.Versions {
			ks.Versions[vid] = val
			byteSize += uint64(len(val))
		}
		byteSize += uint64(len(k.Key))
		next.Keys[k.Key] = ks
	}
	c.byNS[pb.Namespace] = next
	if c.metrics != nil {
		c.metrics.localCacheBytes.set(pb.Namespace, byteSize)
		if businessMoved {
			c.metrics.businessSeqMoved.inc(pb.Namespace)
		}
		if experimentMoved {
			c.metrics.experimentSeqMov.inc(pb.Namespace)
		}
	}
	return true, businessMoved, experimentMoved
}

// knownSeqs returns the per-ns NamespaceSeqs pair the cache currently holds.
// Namespaces with no snapshot are still listed with the zero pair, so the
// server treats a 0/0 entry as "send me everything" per backend §4.1.2.
func (c *configCache) knownSeqs(namespaces []string) map[string]*configv1.NamespaceSeqs {
	out := make(map[string]*configv1.NamespaceSeqs, len(namespaces))
	c.mu.RLock()
	defer c.mu.RUnlock()
	for _, ns := range namespaces {
		s := c.byNS[ns]
		if s == nil {
			out[ns] = &configv1.NamespaceSeqs{}
			continue
		}
		out[ns] = &configv1.NamespaceSeqs{
			BusinessSnapshotSeq:   s.BusinessSnapshotSeq,
			ExperimentSnapshotSeq: s.ExperimentSnapshotSeq,
		}
	}
	return out
}

// listNamespaces returns a sorted snapshot of cached namespace keys, useful
// for diagnostics and tests.
func (c *configCache) listNamespaces() []string {
	c.mu.RLock()
	out := make([]string, 0, len(c.byNS))
	for k := range c.byNS {
		out = append(out, k)
	}
	c.mu.RUnlock()
	sort.Strings(out)
	return out
}
